### PR TITLE
checkers: use cmp.Diff instead of DeepEqual

### PR DESCRIPTION
**Pull request by rogpeppe**
This enables the use of custom equality checkers, and
will panic by default when used on unexported fields which is
something that we should not be doing without explicit
allowance anyway.

Also add CmpEquals to allow the addition of custom equality
checker functionality.

### CODE REVIEW

File: checkers/bool.go [hunk 1]
```diff
@@ -104,14 +104,41 @@ type deepEqualsChecker struct {
 //
 // This checker differs from gocheck.DeepEquals in that
 // it will compare a nil slice equal to an empty slice,
-// and a nil map equal to an empty map.
-var DeepEquals gc.Checker = &deepEqualsChecker{
-	&gc.CheckerInfo{Name: "DeepEquals", Params: []string{"obtained", "expected"}},
-}
+// and a nil map equal to an empty map and that
+// it will panic if called upon to compare unexported fields.
```
Comment (mhilton) [inline checkers/bool.go]: Changing the behavior of unexported fields like this seems a little too drastic a change to me. perhaps leave DeepEquals alone and generally use CmpEquals for new tests.

File: checkers/bool.go [hunk 2]
```diff
@@ -104,14 +104,41 @@ type deepEqualsChecker struct {
...
 func (checker *deepEqualsChecker) Check(params []interface{}, names []string) (result bool, error string) {
 	if ok, err := DeepEqual(params[0], params[1]); !ok {
 		return false, err.Error()
 	}
 	return true, ""
 }
+
+// CmpEquals uses cmp.Diff (see http://godoc.org/github.com/google/go-cmp/cmp#Diff)
+// to compare two values, passing opts to the comparer to enable custom
+// comparison. Passing DeepEqualsOption to CmpEquals makes it
```
Comment (mhilton) [inline checkers/bool.go]: what's DeepEqualsOption?

File: checkers/bool.go [hunk 3]
```diff
@@ -104,14 +104,41 @@ type deepEqualsChecker struct {
...
+		},
+	}
+}
+
+type cmpEqualsChecker struct {
+	*gc.CheckerInfo
+	check func(params []interface{}, names []string) (result bool, error string)
+}
+
+func (c *cmpEqualsChecker) Check(params []interface{}, names []string) (result bool, error string) {
```
Comment (alesstimec) [inline checkers/bool.go]: godoc

**Comment by rogpeppe** [comment]
> I think we should continue to support comparing unexported fields. If you are doing blackbox tests and want to check internal state you should be able to do so. (I don't know tests doing it, but it doesn't feel like it should be forbidden)

It's not forbidden - you just have to enable it explicitly.

There are many tests currently that are doing dubious things like comparing the internal fields of a sync.Mutex. I don't believe this is intentional, and I think that our code would be better off not doing so by default.

The cmp package provides great flexibility in deciding what things to check, and I believe their policy decision to disallow checking of unexported fields by default is the correct one.

> anyway, having cmp.Diff sounds great. Is there something we get from changing DeepEquals?

I'm not sure I understand the question. The new DeepEquals implementation does use cmp.Diff.

**Pull request closed without merging**